timemachine/cpp/src/local_md_potentials.cu

```diff
@@ -0,0 +1,190 @@
+    // Construct a bound potential with 0 params
+    bound_restraint_ = std::shared_ptr<BoundPotential>(new BoundPotential(restraint_, std::vector<int>({0}), nullptr));
+
+    ixn_group_ =
+        construct_ixn_group_potential(N_, nonbonded_bp_->potential, nonbonded_bp_->size(), nonbonded_bp_->d_p->data);
+
+    // Add the restraint potential and ixn group potential
+    all_potentials_.push_back(bound_restraint_);
+    all_potentials_.push_back(ixn_group_);
+
+    cub::DevicePartition::If(
+        nullptr, temp_storage_bytes_, d_free_idxs_.data, d_row_idxs_.data, num_selected_buffer_.data, N_, LessThan(N_));
+    // Allocate char as temp_storage_bytes_ is in raw bytes and the type doesn't matter in practice.
+    // Equivalent to DeviceBuffer<int> buf(temp_storage_bytes_ / sizeof(int))
+    d_temp_storage_buffer_.reset(new DeviceBuffer<char>(temp_storage_bytes_));
+
+    curandErrchk(curandCreateGenerator(&cr_rng_, CURAND_RNG_PSEUDO_DEFAULT));
+};
+
+LocalMDPotentials::~LocalMDPotentials() { curandErrchk(curandDestroyGenerator(cr_rng_)); }
+
+// setup_from_idxs takes a set of idxs, a temperature and a seed to determine the free particles. Fix the local_idxs to length
+// one to ensure the same reference everytime, though the seed also handles the probabilities of selecting particles, and it is suggested
+// to provide a new seed at each step.
+void LocalMDPotentials::setup_from_idxs(
+    double *d_x_t,
+    double *d_box_t,
+    const std::vector<int> &local_idxs,
+    const double temperature,
+    const int seed,
+    const double radius,
+    const double k,
+    const cudaStream_t stream) {
+
+    curandErrchk(curandSetStream(cr_rng_, stream));
+    curandErrchk(curandSetPseudoRandomGeneratorSeed(cr_rng_, seed));
+    // Reset the generator offset to ensure same values for the same seed are produced
+    // Simply reseeding does NOT produce identical results
+    curandErrchk(curandSetGeneratorOffset(cr_rng_, 0));
+
+    // Set the array to all N, which indicates to ignore that idx
+    k_initialize_array<unsigned int><<<ceil_divide(N_, warp_size), warp_size, 0, stream>>>(N_, d_free_idxs_.data, N_);
+    gpuErrchk(cudaPeekAtLastError());
+
+    // Generate values between (0, 1.0]
+    curandErrchk(curandGenerateUniform(cr_rng_, probability_buffer_.data, round_up_even(N_)));
+
+    std::mt19937 rng;
+    rng.seed(seed);
+    std::uniform_int_distribution<unsigned int> random_dist(0, local_idxs.size() - 1);
+
+    unsigned int reference_idx = local_idxs[random_dist(rng)];
+
+    const double kBT = BOLTZ * temperature;
+    // Select all of the particles that will be free
+    k_log_probability_selection<float><<<ceil_divide(N_, warp_size), warp_size, 0, stream>>>(
+        N_, kBT, radius, k, reference_idx, d_x_t, d_box_t, probability_buffer_.data, d_free_idxs_.data);
+    gpuErrchk(cudaPeekAtLastError());
+
+    this->_setup_free(reference_idx, radius, k, stream);
+}
+
+void LocalMDPotentials::_setup_free(
```

Comment:
maybe rename to `_setup_free_idxs()`?

Reply:
Renamed in https://github.com/proteneer/timemachine/pull/1004/commits/23af6f2ae599ddc303f8887155882c3966d6287a